src/ldc/dcompute.d

```diff
@@ -108,4 +114,13 @@ alias SharedPointer(T)   = Pointer!(AddrSpace.Shared,   T);
 alias ConstantPointer(T) = Pointer!(AddrSpace.Constant, immutable(T));
 alias GenericPointer(T)  = Pointer!(AddrSpace.Generic,  T);
 
+// N.B private variables are declared on the stack and so cannot be declared
+// at module scope.
+// No variables exist in the generic address space, it is purely for generalised
+// pointers to point to.
+//
+// The __gshared below does not work. It is kludged into place in `DtoResolveVariable`
 
+alias Global(T)   = __gshared Variable!(AddrSpace.Global,   T);
```

Comment:
Alas this doesn't work the way it looks like it should, and no warning is given!

Excerpt from the test case for the PR to LDC

```d
// LL: @_D{{.*}}bar1{{.*}} = addrspace(1)
Global!float bar1;
// LL: @_D{{.*}}bar1{{.*}} = addrspace(2)
Shared!float bar2;
// LL: @_D{{.*}}bar1{{.*}} = addrspace(3)
Constant!float bar3;
```
Gives
```
%"ldc.dcompute.Variable!(cast(AddrSpace)1u, float).Variable" = type { float }
%"ldc.dcompute.Variable!(cast(AddrSpace)2u, float).Variable" = type { float }
%"ldc.dcompute.Variable!(cast(AddrSpace)3u, float).Variable" = type { float }
@_D22dcompute_cl_addrspaces4bar1S3ldc8dcompute__T8VariableVEQBbQBa9AddrSpacei1TfZQBg = thread_local addrspace(1) global %"ldc.dcompute.Variable!(cast(AddrSpace)1u, float).Variable" { float 0x7FFC000000000000 }, align 4 ; [#uses = 0]
@_D22dcompute_cl_addrspaces4bar2OS3ldc8dcompute__T8VariableVEQBbQBa9AddrSpacei2TfZQBg = addrspace(2) global %"ldc.dcompute.Variable!(cast(AddrSpace)2u, float).Variable" { float 0x7FFC000000000000 }, align 4 ; [#uses = 0]
@_D22dcompute_cl_addrspaces4bar3yS3ldc8dcompute__T8VariableVEQBbQBa9AddrSpacei3TfZQBg = addrspace(3) global %"ldc.dcompute.Variable!(cast(AddrSpace)3u, float).Variable" { float 0x7FFC000000000000 }, align 4 ; [#uses = 0]
```

Reply:
This is totally to be expected in my eyes; whether it's to be stored in TLS or not is not something type-specific IMO. Why don't you use the `shared` type modifier for globals?

Reply:
Because of the way it used. 
`Shared` is thread group local and thus (mostly) requires atomic access and/or synchronisation because it is used for inter thread communication, whereas with `Global` it is read and written with one thread "owning" an index into it so while synchronisation is sometimes necessary it is used much less frequently.